x-pack/plugins/encrypted_saved_objects/server/crypto/encrypted_saved_objects_service.ts

```diff
@@ -94,32 +122,72 @@ export class EncryptedSavedObjectsService {
    * @param type Saved object type.
    */
   public isRegistered(type: string) {
-    return this.typeRegistrations.has(type);
+    return this.typeDefinitions.has(type);
   }
 
   /**
-   * Takes saved object attributes for the specified type and strips any of them that are supposed
-   * to be encrypted and returns that __NEW__ attributes dictionary back.
-   * @param type Type of the saved object to strip encrypted attributes from.
-   * @param attributes Dictionary of __ALL__ saved object attributes.
+   * Takes saved object attributes for the specified type and, depending on the type definition,
+   * either decrypts or strips encrypted attributes (e.g. in case AAD or encryption key has changed
+   * and decryption is no longer possible).
+   * @param descriptor Saved object descriptor (ID, type and optional namespace)
+   * @param attributes Object that includes a dictionary of __ALL__ saved object attributes stored
+   * in Elasticsearch.
+   * @param [originalAttributes] An optional dictionary of __ALL__ saved object original attributes
+   * that were used to create that saved object (i.e. values are NOT encrypted).
+   * @param [params] Parameters that control the way encrypted attributes are handled.
    */
-  public stripEncryptedAttributes<T extends Record<string, unknown>>(
-    type: string,
-    attributes: T
-  ): Record<string, unknown> {
-    const typeRegistration = this.typeRegistrations.get(type);
-    if (typeRegistration === undefined) {
-      return attributes;
+  public async stripOrDecryptAttributes<T extends Record<string, unknown>>(
+    descriptor: SavedObjectDescriptor,
+    attributes: T,
+    originalAttributes?: T,
+    params?: CommonParameters
+  ) {
+    const typeDefinition = this.typeDefinitions.get(descriptor.type);
+    if (typeDefinition === undefined) {
+      return { attributes };
     }
 
+    let decryptedAttributes: T | null = null;
+    let decryptionError: Error | undefined;
     const clonedAttributes: Record<string, unknown> = {};
     for (const [attributeName, attributeValue] of Object.entries(attributes)) {
-      if (!typeRegistration.attributesToEncrypt.has(attributeName)) {
+      // We should strip encrypted attribute if definition explicitly mandates that or decryption
+      // failed.
+      if (typeDefinition.shouldBeStripped(attributeName) || !!decryptionError) {
+        continue;
```

Comment:
While testing, I found that an encryption error prevents plain text attributes from being returned. I think this short-circuit is responsible for that. I would have expected the plain-text attributes to be returned in this scenario

Reply:
I'm blind, thanks for catching this!

Reply:
That means I don't have enough tests as well - I'll fix that

Reply:
Fixed, added unit and integration tests and made sure they fail without fix :)